### DISCUSSION TITLE
Batching/running parallel ecole environment instances

### DISCUSSION THREAD

**Discussion by cwfparsonson** [ds4dm/ml4co-competition#40] [Q&A]
Hi,

Is there any way to batch multiple instances and solve them in parallel using libraries such as `multiprocessing`, `torch.multiprocessing`, or `ray`? I would really like to be able to e.g. generate 32 instances and solve them with my agent in parallel, however because various backend `ecole` objects cannot be pickled I am struggling to do this with the aforementioned libraries. Any advice would be much appreciated!

Best wishes,
Chris

**Comment by gasse** [comment]
Hi @cwfparsonson ,

Ecole natively supports multithreading, so it should be straightforward to add some parallelism to your training algorithm using Python's `[threading`](https://docs.python.org/3/library/threading.html) API. See, e.g., the data collection part in our [dual baseline](https://github.com/ds4dm/ml4co-competition/blob/main/baseline/dual/train_files/01_generate_dataset.py) implementation. Note that although the C/C++ part of the library (which is 99% of Ecole) will run in parallel, the Python part of Ecole and of your code will most likely not benefit from parallelism due to CPython's [GIL](https://wiki.python.org/moin/GlobalInterpreterLock) mechanism.

To get full parallelism with Python's [`multiprocessing`](https://docs.python.org/3/library/multiprocessing.html) API and / or a more high-level library, you will indeed get into trouble if you want to send `ecole` objects through pipes, in particular for classes in `ecole.environment` which are not pickable. I suggest that you only send actions, observations and rewards through pipes, which I think should be [pickable](https://github.com/ds4dm/ecole/pull/152) (if not, let us know). Serializing `ecole.environment` objects is more tricky, as it basically requires to freeze the full current state of the SCIP solver, including part of the registers and stack due to the reverse control mechanism used by Ecole. Also, such objects can rapidly grow in memory as they include the whole B&B tree build by SCIP, so I'm not sure serializing them would be a good practice anyway.

Please feel free to open an issue or a discussion [here](https://github.com/ds4dm/ecole/discussions/new) if you would like to discuss this further with the developers.

Hope this helps,

Best,
Maxime
